### PR TITLE
Async client server

### PR DESCRIPTION
**Pull request by ArtemiyGranat**
TODO:
- [x] Sync client and server compatibility with its async versions
- [x] Async server 
- [x] Async client

Async server:
- [x] Return all client tasks stored in registry to the global queue on failure
- [x] Make async server + sync client tests work
- [x] Add async server + async client tests after async client will be implemented
- [x] Change temporary `malloc` solution for shared client context, probably some global `acl_context_t *` list

Async client:
- [x] Add async client + sync server tests
- [x] Fix bugs

UPD: 
- [x] Valgrind and normal tests with task return mechanism (`netcut`?)
- [ ] Rewrite `recv_wrapper` on `readv`

### CODE REVIEW

File: async_client.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    {
+      print_error ("Could not cancel task queue\n");
+      return (false);
+    }
+  if (queue_cancel (&ctx.result_queue) != QS_SUCCESS)
+    {
+      print_error ("Could not cancel result queue\n");
+      return (false);
+    }
+  if (thread_pool_cancel (&ctx.thread_pool) == S_FAILURE)
```
Comment (hnarkaytis) [inline async_client.c:229]: We need to gracefully collect threads. queue_cancel will initiate graceful shutdown.
change thread_pool_cancel on thread_pool_join

**Review by hnarkaytis — CHANGES_REQUESTED** [review]

File: log.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    {
+      log_level = level;
+      return (S_SUCCESS);
+    }
+
+  if (level < log_level)
+    return (S_SUCCESS);
+
+  char *log;
+  if (asprintf (&log, "(%s %s %d) ", file_name, func_name, line) < 0)
```
Comment (hnarkaytis) [inline log.c]: Let's use statically allocated buffers of some reasonable size. Something like 
char log[1 << 6];
Both for the log[] and for the message[].
Dynamics allocation here is an unnecessary overhead.

File: server_common.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+  return (S_SUCCESS);
+}
+
+status_t
+close_client (int socket_fd)
+{
+  shutdown (socket_fd, SHUT_RDWR);
+  close (socket_fd);
```
Comment (hnarkaytis) [inline server_common.c:84]: Error checks?

File: async_server.c [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *srv_ctx = *(serv_context_t **)arg;
+  mt_context_t *mt_ctx = &srv_ctx->context;
+
+  while (true)
+    {
+      acl_context_t *acl_ctx = acl_context_init (srv_ctx);
```
Comment (hnarkaytis) [inline async_server.c:239]: acl_context_destroy should be in cleanup handler for deallocation in case of thread cancelation

File: async_server.c [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    {
+      error ("Could not allocate memory for client context");
+      return (NULL);
+    }
+  ctx->context = global_ctx;
+
+  if (queue_init (&ctx->registry_idx, sizeof (int)) != QS_SUCCESS)
+    {
+      error ("Could not initialize registry indices queue");
+      return (NULL);
```
Comment (hnarkaytis) [inline async_server.c:33]: Memory leak

File: async_server.c [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    }
+  for (int i = 0; i < QUEUE_SIZE; ++i)
+    if (queue_push (&ctx->registry_idx, &i) != QS_SUCCESS)
+      {
+        error ("Could not push index to registry indices queue");
```
Comment (hnarkaytis) [inline async_server.c:38]: Memory leak

File: async_server.c [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+        return (NULL);
+      }
+
+  ctx->ref_count = 2;
+  if (pthread_mutex_init (&ctx->mutex, NULL) != 0)
+    {
+      error ("Could not initialize mutex");
```
Comment (hnarkaytis) [inline async_server.c:45]: Memory leak

File: async_server.c [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+      trace ("Created a sender thread %08x",
+             mt_ctx->thread_pool.threads.prev->thread);
+
+      if (thread_create (&mt_ctx->thread_pool, result_receiver, &acl_ctx,
+                         sizeof (acl_ctx), "async receiver")
+          == S_FAILURE)
+        {
+          error ("Could not create result receiver thread");
+          close_client (acl_ctx->socket_fd);
```
Comment (hnarkaytis) [inline async_server.c:292]: here we need to decrease ref_count and cancel "sender" thread

File: async_server.c [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+static void
+acl_context_destroy (acl_context_t *ctx)
+{
+  trace ("Destroying client context");
+
+  if (queue_cancel (&ctx->registry_idx) != QS_SUCCESS)
+    {
+      error ("Could not cancel registry indices queue");
+      goto cleanup;
```
Comment (hnarkaytis) [inline async_server.c:60]: we still need to free up queue memory... let's just report errors and continue with normal workflow

File: log.c (hunk 2 above)
Comment (ArtemiyGranat) [reply log.c]: Done

File: server_common.c (hunk 3 above)
Comment (ArtemiyGranat) [reply server_common.c]: Done

File: async_server.c (hunk 4 above)
Comment (ArtemiyGranat) [reply async_server.c:250]: Done

File: async_server.c (hunk 5 above)
Comment (ArtemiyGranat) [reply async_server.c]: Fixed

File: async_server.c (hunk 6 above)
Comment (ArtemiyGranat) [reply async_server.c:38]: Fixed

File: async_server.c (hunk 7 above)
Comment (ArtemiyGranat) [reply async_server.c:45]: Fixed

File: async_server.c (hunk 9 above)
Comment (ArtemiyGranat) [reply async_server.c]: Done

File: async_server.c (hunk 4 above)
Comment (ArtemiyGranat) [reply async_server.c:256]: Done

File: async_server.c (hunk 8 above)
Comment (ArtemiyGranat) [reply async_server.c:310]: Done

File: log.c [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  if (sprintf (log, "(%s %s %d)", file_name, func_name, line) < 0)
+    if (fprintf (stderr, "(%s %s %d)", file_name, func_name, line) < 0)
+      return (S_FAILURE);
+
+  va_list args;
+  va_start (args, msg);
+
+  char message[1 << 7];
+  if (vsprintf (message, msg, args) < 0)
+    if (vfprintf (stderr, msg, args) < 0)
```
Comment (hnarkaytis) [inline log.c:24]: better to use snprintf

File: log.c [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#include "log.h"
+#include "common.h"
+
+#include <pthread.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+__attribute__ ((format (printf, 4, 5))) status_t
+message_impl (const char *file_name, const char *func_name, int line,
```
Comment (hnarkaytis) [inline log.c:11]: Let's add log level to the message

File: async_server.c [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+
+static void
+sender_receiver_cleanup (void *arg)
+{
+  acl_context_t *ctx = arg;
+
+  bool is_last;
+  if (pthread_mutex_lock (&ctx->mutex) != 0)
+    return;
+  pthread_cleanup_push (cleanup_mutex_unlock, &ctx->mutex);
```
Comment (hnarkaytis) [inline async_server.c:122]: I'm still hesitant about return_tasks under lock.

As an alternative aproach, we could store receiver/sender threads ips in client context and explicitly thread_cancel both. Then return_tasks could be moved to if (is_last) branch after lock.

File: async_server.c [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+        }
+
+      trace ("Pushed index of received task back to indices queue");
+
+      cl_ctx->registry_used[task.id] = false;
+
+      if (serv_signal_if_found (mt_ctx) == S_FAILURE)
+        break;
+
+      if (mt_ctx->password[0] != 0)
```
Comment (hnarkaytis) [inline async_server.c:176]: Is it really required?

File: log.c (hunk 10 above)
Comment (ArtemiyGranat) [reply log.c]: Fixed

File: log.c (hunk 11 above)
Comment (ArtemiyGranat) [reply log.c]: Done

File: async_server.c (hunk 13 above)
Comment (ArtemiyGranat) [reply async_server.c]: Somewhy I thought that it is but its not required

File: async_server.c (hunk 12 above)
Comment (ArtemiyGranat) [reply async_server.c:122]: I added a TODO mark to try to do this with a fresh head after I do some tasks other than async client-server, so can we squash & merge now?

File: log.h [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+#define FATAL_OFF IGNORE_MESSAGE
+
+#define PASTE2(...) PASTE2_ (__VA_ARGS__)
+#define PASTE2_(_0, _1) _0##_##_1
+
+#ifndef LOG_LEVEL
+#define LOG_LEVEL ERROR
+#endif
+
+#define trace(...) PASTE2 (TRACE, LOG_LEVEL) (__VA_ARGS__)
```
Comment (hnarkaytis) [inline log.h:72]: ``It's better to pass log_level as enum here:
`#define trace(...) PASTE2 (TRACE, LOG_LEVEL) (LL_TRACE, __VA_ARGS__)`

In message_impl you will need a mapping from enum to string:
```
#define LL_ENTRY(LEVEL) [LL_##LEVEL] = #LEVEL,
log_level_str[] = { LL_ENTRY (TRACE), LL_ENTRY (DEBUG), ...};
```

Comment (ArtemiyGranat) [reply log.h]: Done

**Review by hnarkaytis — APPROVED** [review]
squash and merge

**Pull request merged**